packages/dds/tree/src/feature-libraries/persisted-config/persistedConfig.ts

```diff
@@ -0,0 +1,336 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+
+// TODO: Document, decide if semver-like even provides value here (over e.g. plain-old number).
+export type Version = `${number}.${number}.${number}${"-internal" | ""}`;
```

Comment:
I thought similarly at first, but after further thinking on the problem I'm not sure this is a valuable way of thinking about it. For one thing, it doesn't communicate the difference between a format change which just reencodes the same information and one which invalidates in-flight ops.

For another, under that definition I generally don't think applications would want to roll out "minor updates" until they've ensured the code which understands the newer format has saturated: if you're adding a new optional field to the persisted format, that field will generally have some meaning. For all of the examples I thought of, it would be problematic if code not understanding the optional field began collaborating with code that did.

One example of this (though it's not in FormatVersion, it would be in the config "feature flags") is attribution. TL;DR of that is that there is no change whatsoever to the op format, but it's expected clients which have attribution on store more information about the ops that are coming in (you're somewhat familiar with the design; its overall shape hasn't changed much since you reviewed). If a client that doesn't understand attribution starts collaborating with one that does and that client becomes the summarizer, all of the attribution data on the document may be lost.

I think basically the only kind of change it's safe to make to persisted format (op, summary) without caring about code saturation are "normalization" ones, where you combine two equivalent representations of the same data in the old format. Examples of this include not writing empty traits, or recent change Jillian made to merge-tree which combined empty/undefined property sets.

Given that applications therefore want formats to generally saturate before turning them on for both major and minor versions, it seems arbitrary to draw some line between the types of changes. My more recent thinking on this has been that a number is probably sufficient, and the only real cases we want to distinguish are changes which invalidate in-flight ops from those that don't. I'm not sure using semver for that would be natural.

Reply:
I also ended up thinking a single number is enough last time I worked through this.